invenio_search/api.py

```diff
@@ -117,3 +119,48 @@ def search(self):
                 return search_.response_class(partial(FacetedResponse, self))
 
         return RecordsFacetedSearch(query=query, filters=filters or {})
+
+    def with_preference_param(self):
+        """Add the preference param to the ES request and return a new Search.
+
+        The preference param avoids the bouncing effect with multiple
+        replicas, documented on ES documentation.
+        See: https://www.elastic.co/guide/en/elasticsearch/guide/current
+        /_search_options.html#_preference for more information.
+        """
+        user_hash = self._get_user_hash()
+        if user_hash:
+            return self.params(preference=user_hash)
+        return self
+
+    def _get_user_agent(self):
+        """Retrieve the request's User-Agent, if available.
+
+        Taken from Flask Login utils.py.
+        """
+        user_agent = request.headers.get('User-Agent')
+        if user_agent is not None:
+            user_agent = user_agent.encode('utf-8')
+        return user_agent or ''
+
+    def _get_remote_addr(self):
+        """Retrieve the request's address ip, if available.
+
+        Taken from Flask Login utils.py.
+        """
+        address = request.headers.get('X-Forwarded-For', request.remote_addr)
```

Comment:
You can just use ``request.remote_addr``, since we expect that to be set correctly. In case of proxies, you need to use Werkzeug's WSGIProxyFix to set it correctly from X-Forwarded-For headers. In order to correctly determine the right user IP address, you need to know how many proxies are in-front of your webapp. For Invenio this is usually 2: Nginx + HAProxy, hence, making the users' IP appear as second in the list. :-)